File: client/client-multi/client-objective-c/Kaa/channel/impl/channels/AbstractHttpChannel.m
/*
 * Copyright 2014-2016 CyberVision, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#import "AbstractHttpChannel.h"
#import "IPTransportInfo.h"
#import "TransportProtocolIdHolder.h"
#import "KaaLogging.h"

static NSString *const logTag = @"AbstractHttpChannel >>>";

typedef NS_ENUM(int, HttpStatus) {
    HttpStatusUnauthorised = 401,
    HttpStatusForbidden = 403
};

@interface AbstractHttpChannel ()

@property (nonatomic, strong) IPTransportInfo *currentServer;
@property (nonatomic, weak) AbstractKaaClient *kaaClient;
@property (nonatomic, strong) id<KaaClientState> kaaState;
@property (nonatomic, strong) id<FailoverManager> failoverMgr;

@property (nonatomic, strong) volatile NSOperationQueue *executor;

@property (nonatomic) volatile BOOL lastConnectionFailed;
@property (nonatomic) volatile BOOL isPaused;

@property (nonatomic, strong) AbstractHttpClient *kaaHttpClient;
@property (nonatomic, strong) id<KaaDataDemultiplexer> chDemultiplexer;
@property (nonatomic, strong) id<KaaDataMultiplexer> chMultiplexer;

@end

@implementation AbstractHttpChannel

- (instancetype)initWithClient:(AbstractKaaClient *)client
                         state:(id<KaaClientState>)state
               failoverManager:(id<FailoverManager>)manager {
    self = [super init];
    if (self) {
        self.kaaClient = client;
        self.kaaState = state;
        self.failoverMgr = manager;
    }
    return self;
}

- (TransportProtocolId *)getTransportProtocolId {
    return [TransportProtocolIdHolder HTTPTransportID];
}

- (void)syncForTransportType:(TransportType)type {
    @synchronized(self) {
        [self syncTransportTypes:[NSSet setWithObject:@(type)]];
    }
}

- (void)syncTransportTypes:(NSSet *)types {
    @synchronized(self) {
        if (self.isShutdown) {
            DDLogInfo(@"%@ Can't sync. Channel %@ is down", logTag, [self getId]);
            return;
        }
        if (self.isPaused) {
            DDLogInfo(@"%@ Can't sync. Channel %@ is paused", logTag, [self getId]);
            return;
        }
        if (!self.chMultiplexer) {
            DDLogInfo(@"%@ Can't sync. Channel %@ multiplexer is not set", logTag, [self getId]);
            return;
        }
        if (!self.chDemultiplexer) {
            DDLogWarn(@"%@ Can't sync. Channel %@ demultiplexer is not set", logTag, [self getId]);
            return;
        }
        if (!self.currentServer) {
            self.lastConnectionFailed = YES;
            DDLogWarn(@"%@ Can't sync. Server is nil", logTag);
        }
        
        NSMutableDictionary *typeMap = [NSMutableDictionary dictionary];
        for (NSNumber *type in types) {
            DDLogInfo(@"%@ Processing sync %i for channel %@", logTag, [type intValue], [self getId]);
            NSNumber *channelDirection = [self getSupportedTransportTypes][type];
            if (channelDirection) {
                typeMap[type] = channelDirection;
            } else {
                DDLogError(@"%@ Unsupported type %i for channel %@", logTag, [type intValue], [self getId]);
            }
        }
        if (self.executor) {
            __weak typeof(self) weakSelf = self;
            [self.executor addOperationWithBlock:^{
                [weakSelf processTypes:typeMap];
            }];
        } else {
            DDLogError(@"%@ No executor found for channel with id: %@", logTag, [self getId]);
        }
    }
}

- (void)syncAll {
    @synchronized(self) {
        if (self.isShutdown) {
            DDLogInfo(@"%@ Can't sync all. Channel %@ is down", logTag, [self getId]);
            return;
        }
        if (self.isPaused) {
            DDLogInfo(@"%@ Can't sync. Channel %@ is paused", logTag, [self getId]);
            return;
        }
        
        if (!self.chMultiplexer || !self.chDemultiplexer) {
            DDLogWarn(@"%@ Can't sync, multiplexer/demultiplexer not set: %@/%@", logTag, self.chMultiplexer, self.chDemultiplexer);
            return;
        }
        if (self.currentServer) {
            __weak typeof(self) weakSelf = self;
            [self.executor addOperationWithBlock:^{
                [weakSelf processTypes:[weakSelf getSupportedTransportTypes]];
            }];
        } else {
            self.lastConnectionFailed = YES;
            DDLogWarn(@"%@ Can't sync. Server is nil", logTag);
        }
    }
}

- (void)syncAckForTransportType:(TransportType)type {
    [self syncAckForTransportTypes:[NSSet setWithObject:@(type)]];
}

- (void)syncAckForTransportTypes:(NSSet *)types {
#pragma unused(types)
    DDLogInfo(@"%@ Sync ack message is ignored for Channel with id: %@", logTag, [self getId]);
}

- (void)setMultiplexer:(id<KaaDataMultiplexer>)multiplexer {
    @synchronized(self) {
        if (multiplexer) {
            self.chMultiplexer = multiplexer;
        }
    }
}

- (void)setDemultiplexer:(id<KaaDataDemultiplexer>)demultiplexer {
    @synchronized(self) {
        if (demultiplexer) {
            self.chDemultiplexer = demultiplexer;
        }
    }
}

- (void)setServer:(id<TransportConnectionInfo>)server {
    @synchronized(self) {
        if (self.isShutdown) {
            DDLogInfo(@"%@ Can't set server. Channel %@ is down", logTag, [self getId]);
            return;
        }
        if (!self.executor && !self.isPaused) {
            self.executor = [self createExecutor];
        }
        if (server) {
            self.currentServer = [[IPTransportInfo alloc] initWithTransportInfo:server];
            NSString *url = [NSString stringWithFormat:@"%@%@", [self.currentServer getUrl], [self getURLSuffix]];
            self.kaaHttpClient = [self.kaaClient createHttpClientWithURLString:url
                                                                 privateKeyRef:[self.kaaState privateKey]
                                                                  publicKeyRef:[self.kaaState publicKey]
                                                                     remoteKey:[self.currentServer getPublicKey]];
            if (self.lastConnectionFailed && !self.isPaused) {
                self.lastConnectionFailed = NO;
                [self syncAll];
            }
        }
    }
}

- (id<TransportConnectionInfo>)getServer {
    return self.currentServer;
}

- (void)setConnectivityChecker:(ConnectivityChecker *)checker {
#pragma unused (checker)
    DDLogInfo(@"%@ Ignore set connectivity checker", logTag);
}

- (void)shutdown {
    if (!self.isShutdown) {
        _isShutdown = YES;
        if (self.executor) {
            [self.executor cancelAllOperations];
        }
    }
}

- (void)pause {
    if (self.isShutdown) {
        DDLogInfo(@"%@ Can't pause. Channel %@ is down", logTag, [self getId]);
        return;
    }
    if (!self.isPaused) {
        self.isPaused = YES;
        if (self.executor) {
            [self.executor cancelAllOperations];
            self.executor = nil;
        }
    }
}

- (void)resume {
    if (self.isShutdown) {
        DDLogInfo(@"%@ Can't resume. Channel %@ is down", logTag, [self getId]);
        return;
    }
    
    if (self.isPaused) {
        self.isPaused = NO;
        if (!self.executor) {
            self.executor = [self createExecutor];
        }
        if (self.lastConnectionFailed) {
            self.lastConnectionFailed = NO;
            [self syncAll];
        }
    }
}

- (ServerType)getServerType {
    [NSException raise:NSInternalInconsistencyException format:@"Not implemented in abstract class"];
    return -1;
}

- (NSDictionary *)getSupportedTransportTypes {
    [NSException raise:NSInternalInconsistencyException format:@"Not implemented in abstract class"];
    return nil;
}

- (NSString *)getId {
    [NSException raise:NSInternalInconsistencyException format:@"Not implemented in abstract class"];
    return nil;
}

- (void)connectionEstablished {
    self.lastConnectionFailed = NO;
    [self.failoverMgr onServerConnectedWithConnectionInfo:self.currentServer];
}

- (void)connectionFailedWithStatus:(int)status {
    FailoverStatus failoverStatus = FailoverStatusOperationsServersNotAvailable;
    switch (status) {
        case HttpStatusUnauthorised:
            [self.kaaState clean];
            failoverStatus = FailoverStatusEndpointVerificationFailed;
            break;
        case HttpStatusForbidden:
            failoverStatus = FailoverStatusEndpointCredentialsRevoked;
            break;
        default:
            break;
    }
    self.lastConnectionFailed = YES;
    [self.failoverMgr onServerFailedWithConnectionInfo:self.currentServer failoverStatus:failoverStatus];
}

- (id<KaaDataMultiplexer>)getMultiplexer {
    return self.chMultiplexer;
}

- (id<KaaDataDemultiplexer>)getDemultiplexer {
    return self.chDemultiplexer;
}

- (AbstractHttpClient *)getHttpClient {
    return self.kaaHttpClient;
}

- (void)processTypes:(NSDictionary *)types {
#pragma unused(types)
    [NSException raise:NSInternalInconsistencyException format:@"Not implemented in abstract class"];
}

- (NSString *)getURLSuffix {
    [NSException raise:NSInternalInconsistencyException format:@"Not implemented in abstract class"];
    return nil;
}

- (NSOperationQueue *)createExecutor {
    DDLogInfo(@"%@ Creating a new executor for channel: %@", logTag, [self getId]);
    NSOperationQueue *queue = [[NSOperationQueue alloc] init];
    queue.maxConcurrentOperationCount = 1;
    return queue;
}

@end

File: client/client-multi/client-java-core/src/main/java/org/kaaproject/kaa/client/channel/impl/channels/AbstractHttpChannel.java
/*
 * Copyright 2014-2016 CyberVision, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.kaaproject.kaa.client.channel.impl.channels;

import org.kaaproject.kaa.client.AbstractKaaClient;
import org.kaaproject.kaa.client.channel.ChannelDirection;
import org.kaaproject.kaa.client.channel.IpTransportInfo;
import org.kaaproject.kaa.client.channel.KaaDataChannel;
import org.kaaproject.kaa.client.channel.KaaDataDemultiplexer;
import org.kaaproject.kaa.client.channel.KaaDataMultiplexer;
import org.kaaproject.kaa.client.channel.TransportConnectionInfo;
import org.kaaproject.kaa.client.channel.TransportProtocolId;
import org.kaaproject.kaa.client.channel.TransportProtocolIdConstants;
import org.kaaproject.kaa.client.channel.connectivity.ConnectivityChecker;
import org.kaaproject.kaa.client.channel.failover.FailoverManager;
import org.kaaproject.kaa.client.channel.failover.FailoverStatus;
import org.kaaproject.kaa.client.persistence.KaaClientState;
import org.kaaproject.kaa.client.transport.AbstractHttpClient;
import org.kaaproject.kaa.common.TransportType;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.Collections;
import java.util.HashMap;
import java.util.Map;
import java.util.Set;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

public abstract class AbstractHttpChannel implements KaaDataChannel {
  public static final Logger LOG = LoggerFactory // NOSONAR
      .getLogger(AbstractHttpChannel.class);

  private static final int UNAUTHORIZED_HTTP_STATUS = 401;
  private static final int FORBIDDEN_HTTP_STATUS = 403;
  private final AbstractKaaClient client;
  private final KaaClientState state;
  private final FailoverManager failoverManager;
  private IpTransportInfo currentServer;
  private volatile ExecutorService executor;

  private volatile boolean lastConnectionFailed = false;
  private volatile boolean isShutdown = false;
  private volatile boolean isPaused = false;

  private AbstractHttpClient httpClient;
  private KaaDataDemultiplexer demultiplexer;
  private KaaDataMultiplexer multiplexer;

  /**
   * All-args constructor.
   */
  public AbstractHttpChannel(AbstractKaaClient client, KaaClientState state,
                             FailoverManager failoverManager) {
    this.client = client;
    this.state = state;
    this.failoverManager = failoverManager;
  }

  protected ExecutorService createExecutor() {
    LOG.info("Creating a new executor for channel {}", getId());
    return Executors.newSingleThreadExecutor();
  }

  @Override
  public TransportProtocolId getTransportProtocolId() {
    return TransportProtocolIdConstants.HTTP_TRANSPORT_ID;
  }

  @Override
  public synchronized void sync(TransportType type) {
    sync(Collections.singleton(type));
  }

  @Override
  public synchronized void sync(Set<TransportType> types) {
    if (isShutdown) {
      LOG.info("Can't sync. Channel {} is down", getId());
      return;
    }
    if (isPaused) {
      LOG.info("Can't sync. Channel {} is paused", getId());
      return;
    }
    if (multiplexer == null) {
      LOG.warn("Can't sync. Channel {} multiplexer is not set", getId());
      return;
    }
    if (demultiplexer == null) {
      LOG.warn("Can't sync. Channel {} demultiplexer is not set", getId());
      return;
    }
    if (currentServer == null) {
      lastConnectionFailed = true;
      LOG.warn("Can't sync. Server is null");
    }
    Map<TransportType, ChannelDirection> typeMap = new HashMap<>(1);
    for (TransportType type : types) {
      LOG.info("Processing sync {} for channel {}", type, getId());
      ChannelDirection direction = getSupportedTransportTypes().get(type);
      if (direction != null) {
        typeMap.put(type, direction);
      } else {
        LOG.error("Unsupported type {} for channel {}", type, getId());
      }
    }
    executor.submit(createChannelRunnable(typeMap));
  }

  @Override
  public synchronized void syncAll() {
    if (isShutdown) {
      LOG.info("Can't sync. Channel {} is down", getId());
      return;
    }
    if (isPaused) {
      LOG.info("Can't sync. Channel {} is paused", getId());
      return;
    }
    LOG.info("Processing sync all for channel {}", getId());
    if (multiplexer != null && demultiplexer != null) {
      if (currentServer != null) {
        executor.submit(createChannelRunnable(getSupportedTransportTypes()));
      } else {
        lastConnectionFailed = true;
        LOG.warn("Can't sync. Server is null");
      }
    }
  }

  @Override
  public void syncAck(TransportType type) {
    syncAck(Collections.singleton(type));
  }

  @Override
  public void syncAck(Set<TransportType> types) {
    LOG.info("Sync ack message is ignored for Channel {}", getId());
  }

  protected abstract String getUrlSufix();

  @Override
  public TransportConnectionInfo getServer() {
    return currentServer;
  }

  @Override
  public synchronized void setServer(TransportConnectionInfo server) {
    if (isShutdown) {
      LOG.info("Can't set server. Channel {} is down", getId());
      return;
    }
    if (executor == null && !isPaused) {
      executor = createExecutor();
    }
    if (server != null) {
      this.currentServer = new IpTransportInfo(server);
      this.httpClient = client.createHttpClient(currentServer.getUrl() + getUrlSufix(),
              state.getPrivateKey(), state.getPublicKey(),
          currentServer.getPublicKey());
      if (lastConnectionFailed && !isPaused) {
        lastConnectionFailed = false;
        syncAll();
      }
    }
  }

  @Override
  public void setConnectivityChecker(ConnectivityChecker checker) {
  }

  @Override
  public void shutdown() {
    if (!isShutdown) {
      isShutdown = true;
      if (executor != null) {
        executor.shutdownNow();
      }
    }
  }

  @Override
  public void pause() {
    if (isShutdown) {
      LOG.info("Can't pause channel. Channel [{}] is down", getId());
      return;
    }
    if (!isPaused) {
      isPaused = true;
      if (executor != null) {
        executor.shutdownNow();
        executor = null;
      }
    }
  }

  @Override
  public void resume() {
    if (isShutdown) {
      LOG.info("Can't resume channel. Channel [{}] is down", getId());
      return;
    }
    if (isPaused) {
      isPaused = false;
      if (executor == null) {
        executor = createExecutor();
      }
      if (lastConnectionFailed) {
        lastConnectionFailed = false;
        syncAll();
      }
    }
  }

  protected void connectionFailed(boolean failed) {
    connectionFailed(failed, -1);
  }

  protected void connectionFailed(boolean failed, int status) {
    FailoverStatus failoverStatus = FailoverStatus.OPERATION_SERVERS_NA;
    switch (status) {
      case UNAUTHORIZED_HTTP_STATUS:
        state.clean();
        failoverStatus = FailoverStatus.ENDPOINT_VERIFICATION_FAILED;
        break;
      case FORBIDDEN_HTTP_STATUS:
        failoverStatus = FailoverStatus.ENDPOINT_CREDENTIALS_REVOKED;
        break;
      default:
        break;
    }

    lastConnectionFailed = failed;
    if (failed) {
      failoverManager.onServerFailed(currentServer, failoverStatus);
    } else {
      failoverManager.onServerConnected(currentServer);
    }
  }

  protected KaaDataMultiplexer getMultiplexer() {
    return multiplexer;
  }

  @Override
  public synchronized void setMultiplexer(KaaDataMultiplexer multiplexer) {
    if (multiplexer != null) {
      this.multiplexer = multiplexer;
    }
  }

  protected KaaDataDemultiplexer getDemultiplexer() {
    return demultiplexer;
  }

  @Override
  public synchronized void setDemultiplexer(KaaDataDemultiplexer demultiplexer) {
    if (demultiplexer != null) {
      this.demultiplexer = demultiplexer;
    }
  }

  protected AbstractHttpClient getHttpClient() {
    return httpClient;
  }

  protected abstract Runnable createChannelRunnable(Map<TransportType, ChannelDirection> typeMap);

  public boolean isShutdown() {
    return isShutdown;
  }
}
